### PR TITLE
Add recursive structs supported

### PR DESCRIPTION
**Pull request by RyougiNevermore**
just use processing cache, and no performance loss.

### CODE REVIEW

File: codec_record.go [hunk 1]
```diff
@@ -54,6 +54,13 @@ func createEncoderOfRecord(cfg *frozenConfig, schema Schema, typ reflect2.Type)
 }
 
 func decoderOfStruct(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValDecoder {
+	processing := cfg.getProcessingDecoderFromCache(schema.Fingerprint(), typ.RType())
+	if processing != nil {
+		return processing
+	}
+	dec := &structDecoder{typ: typ, fields: nil}
+	cfg.addProcessingDecoderToCache(schema.Fingerprint(), typ.RType(), dec)
```
Comment (nrwiersma) [inline codec_record.go:62]: This sets up a race in parallel processing that will lead to SIGSEGV.

**Comment by RyougiNevermore** [comment]
I'm not sure about the results of golangci-lint. any help?

**Comment by RyougiNevermore** [comment]
uhhh, if support disable cache mode, maybe use some one like ctx, and it will cost more performance, from 1400 ns/op up to 1700 ns/op in.
and there is a bug about big.rat in generic decoder..., so can use *big.rat insteadof big.rat?

File: processings.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+func newProcessingGroup() *processingGroup {
+	return &processingGroup{
+		group: new(singleflight.Group),
+		pool:  new(sync.Pool),
+	}
+}
+
+type processingGroup struct {
+	group *singleflight.Group
```
Comment (nrwiersma) [inline processings.go:71]: This is not needed for the local option.

File: processings.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	"golang.org/x/sync/singleflight"
+	"sync"
+	"unsafe"
+)
+
+type decoderOfTypeHandler func(cfg *frozenConfig, p *processing, schema Schema, typ reflect2.Type) ValDecoder
+type encoderOfTypeHandler func(cfg *frozenConfig, p *processing, schema Schema, typ reflect2.Type) ValEncoder
+
+type processing struct {
+	key      []byte
```
Comment (nrwiersma) [inline processings.go:15]: I dont get what this is used for?

File: internal/mmhash/hash.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// Package mmhash export runtime.memhash
+package mmhash
```
Comment (nrwiersma) [inline internal/mmhash/hash.go:2]: This seems to be unused.

**Review by nrwiersma — CHANGES_REQUESTED** [review]
Please rebase and clean up. Seems a lot of stuff is left over or unused.

**Comment by nrwiersma** [comment]
Fixed by #413

**Pull request closed without merging**